### PR TITLE
PicoTTS added - compiled from source

### PR DESCRIPTION
**Pull request by carstencodes**
Enhancemend for #135 by @kngharv: Building PicoTTS from source.

### CODE REVIEW

**Comment by homeassistant** [comment]
Hi @carstencodes, @kngharv, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;docker#142).

Once you do that we will be able to review and accept this pull request.

Thanks!

File: base/Dockerfile [hunk 1]
```diff
@@ -100,6 +100,18 @@ RUN apk add --no-cache \
     && apk del .build-dependencies \
     && rm -rf /usr/src/arp-scan
 
+# PicoTTS - it has no specific version - unfortunately we'll have to use master directly
+RUN apk add automake autoconf git libtool popt-dev git build-base \ 
+    && mkdir -p -m 755 /usr/src/pico \
```
Comment (frenck) [inline base/Dockerfile:105]: This folder is created by the git clone below.

File: base/Dockerfile [hunk 2]
```diff
@@ -100,6 +100,18 @@ RUN apk add --no-cache \
     && apk del .build-dependencies \
     && rm -rf /usr/src/arp-scan
 
+# PicoTTS - it has no specific version - unfortunately we'll have to use master directly
+RUN apk add automake autoconf git libtool popt-dev git build-base \ 
+    && mkdir -p -m 755 /usr/src/pico \
+    && git clone https://github.com/naggety/picotts.git /usr/src/pico \
```
Comment (frenck) [inline base/Dockerfile:106]: The working directory is already `/usr/src`, so just `pico` will do.

```suggestion
    && git clone https://github.com/naggety/picotts.git pico \

```

File: base/Dockerfile [hunk 3]
```diff
@@ -100,6 +100,18 @@ RUN apk add --no-cache \
     && apk del .build-dependencies \
     && rm -rf /usr/src/arp-scan
 
+# PicoTTS - it has no specific version - unfortunately we'll have to use master directly
+RUN apk add automake autoconf git libtool popt-dev git build-base \ 
+    && mkdir -p -m 755 /usr/src/pico \
+    && git clone https://github.com/naggety/picotts.git /usr/src/pico \
+    && cd /usr/src/pico/pico \
```
Comment (frenck) [inline base/Dockerfile:107]:
```suggestion
    && cd pico \
```

File: base/Dockerfile [hunk 4]
```diff
@@ -100,6 +100,18 @@ RUN apk add --no-cache \
...
+RUN apk add automake autoconf git libtool popt-dev git build-base \ 
+    && mkdir -p -m 755 /usr/src/pico \
+    && git clone https://github.com/naggety/picotts.git /usr/src/pico \
+    && cd /usr/src/pico/pico \
+    && ./autogen.sh \
+    && ./configure \
+    && make \
+    && make install \
+    && cd - \
+    && rm -rfv /usr/src/pico
```
Comment (frenck) [inline base/Dockerfile:113]: No need to be verbose here

```suggestion
    && rm -rf /usr/src/pico
```

File: base/Dockerfile [hunk 5]
```diff
@@ -100,6 +100,18 @@ RUN apk add --no-cache \
     && apk del .build-dependencies \
     && rm -rf /usr/src/arp-scan
 
+# PicoTTS - it has no specific version - unfortunately we'll have to use master directly
+RUN apk add automake autoconf git libtool popt-dev git build-base \
```
Comment (frenck) [inline base/Dockerfile:104]: These are build dependencies, they need to be installed virtual and cleaned up after the build has finished. Also, please note, we don't have any package cache, so you'll need to use `--no-cache` as well.

**Review by frenck — CHANGES_REQUESTED** [review]

File: base/Dockerfile [hunk 6]
```diff
@@ -100,6 +100,18 @@ RUN apk add --no-cache \
     && apk del .build-dependencies \
     && rm -rf /usr/src/arp-scan
 
+# PicoTTS - it has no specific version - unfortunately we'll have to use master directly
```
Comment (frenck) [inline base/Dockerfile:103]: We want reproducible builds, please lock it down to commit hash (as an alternative to version tags).

**Comment by pvizeli** [comment]
Please post image size differents

**Comment by homeassistant** [comment]
Hi @kngharv, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;docker#142).

Once you do that we will be able to review and accept this pull request.

Thanks!

File: base/Dockerfile (hunk 1 above)
Comment (carstencodes) [reply base/Dockerfile]: The update should solve this.

File: base/Dockerfile (hunk 3 above)
Comment (carstencodes) [reply base/Dockerfile]: Solved in the update

File: base/Dockerfile (hunk 5 above)
Comment (carstencodes) [reply base/Dockerfile]: Added argument using the full commit hash.

File: base/Dockerfile (hunk 4 above)
Comment (carstencodes) [reply base/Dockerfile]: Fixed in the update

File: base/Dockerfile (hunk 2 above)
Comment (carstencodes) [reply base/Dockerfile]: Fixed in the update

File: base/Dockerfile (hunk 6 above)
Comment (carstencodes) [reply base/Dockerfile]: Added commit hash as argument

**Comment by carstencodes** [comment]
> Please post image size differents

@pvizeli About 11 MB more on a pure alpine image.

File: base/Dockerfile [hunk 7]
```diff
@@ -100,17 +100,20 @@ RUN apk add --no-cache \
     && apk del .build-dependencies \
     && rm -rf /usr/src/arp-scan
 
-# PicoTTS - it has no specific version - unfortunately we'll have to use master directly
-RUN apk add automake autoconf git libtool popt-dev git build-base \ 
-    && mkdir -p -m 755 /usr/src/pico \
-    && git clone https://github.com/naggety/picotts.git /usr/src/pico \
-    && cd /usr/src/pico/pico \
+# PicoTTS - it has no specific version - use current master branch commit
+ARG PICOTTS_HASH=e3ba46009ee868911fa0b53db672a55f9cc13b1c
```
Comment (pvizeli) [inline base/Dockerfile:104]: Push this into `build.json`

File: base/Dockerfile [hunk 8]
```diff
@@ -100,17 +100,20 @@ RUN apk add --no-cache \
...
+    && cd pico \ # Sources and make files are in sub-dir ./pico
     && ./autogen.sh \
     && ./configure \
     && make \
     && make install \
-    && cd - \
-    && rm -rfv /usr/src/pico
+    && cd ../.. \
+    && rm -rf /usr/src/pico \
+    && apk del .build-dependencies
```
Comment (pvizeli) [inline base/Dockerfile:116]: follow the style from Telldus + you don't need to take out of the folder if you remove it at the latest.

**Review by pvizeli — CHANGES_REQUESTED** [review]

**Comment by homeassistant** [comment]
Hi @kngharv, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;docker#142).

Once you do that we will be able to review and accept this pull request.

Thanks!

File: base/Dockerfile (hunk 8 above)
Comment (carstencodes) [reply base/Dockerfile]: I followed it. It should work. See 2nd update.

File: base/Dockerfile (hunk 7 above)
Comment (carstencodes) [reply base/Dockerfile]: Never tried to do it that way. Hope, it has been done correclty, could not test it.

I took my orientation from the arguments above. Hope it works.

File: base/Dockerfile [hunk 9]
```diff
@@ -100,6 +100,20 @@ RUN apk add --no-cache \
     && apk del .build-dependencies \
     && rm -rf /usr/src/arp-scan
 
+# PicoTTS - it has no specific version - commit should be taken from build.json
+ARG PICOTTS_HASH
+RUN apk add --no-cache --virtual .build-dependencies automake autoconf libtool popt-dev build-base \
```
Comment (pvizeli) [inline base/Dockerfile:105]: `popt-dev` your compiled software will not link against this library?

File: base/build.json [hunk 10]
```diff
@@ -1,5 +1,6 @@
 {
   "image": "homeassistant/{arch}-homeassistant-base",
+  "picotts_hash": "e3ba46009ee868911fa0b53db672a55f9cc13b1c",
```
Comment (pvizeli) [inline base/build.json:3]: An example how it should look like: https://github.com/home-assistant/hassio-addons/blob/master/zwave/build.json
 - Looking also into https://github.com/home-assistant/hassio-addons/blob/master/zwave/Dockerfile to understand how the Dockerfile should be formated

**Review by pvizeli — CHANGES_REQUESTED** [review]

**Comment by pvizeli** [comment]
Rebase also your branch again master to get the latest changes. Our CI will show you now if your work can be built or not

**Comment by homeassistant** [comment]
Hi @kngharv, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;docker#142).

Once you do that we will be able to review and accept this pull request.

Thanks!

File: base/Dockerfile (hunk 9 above)
Comment (carstencodes) [reply base/Dockerfile]: Fixed with latest commit after rebase

Comment (carstencodes) [reply base/Dockerfile]: Thought that popt uses static-linking in this case, but I looked it up and you're right.

File: base/build.json (hunk 10 above)
Comment (carstencodes) [reply base/build.json]: Rebasing took place. Update and CI/CD in progress.

Thanks for the advice.

**Comment by homeassistant** [comment]
Hi @kngharv, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;docker#142).

Once you do that we will be able to review and accept this pull request.

Thanks!

**Comment by pvizeli** [comment]
https://github.com/home-assistant/docker/pull/142/checks?check_run_id=1310419996#step:5:117 looks wrong. There is no cleanup

File: Dockerfile [hunk 11]
```diff
@@ -100,6 +100,21 @@ RUN apk add --no-cache \
     && apk del .build-dependencies \
     && rm -rf /usr/src/arp-scan
 
+# PicoTTS - it has no specific version - commit should be taken from build.json
+ARG PICOTTS_HASH
+RUN apk add --no-cache popt \
```
Comment (pvizeli) [inline Dockerfile:105]: packages on a new line

File: Dockerfile [hunk 12]
```diff
@@ -100,6 +100,21 @@ RUN apk add --no-cache \
     && apk del .build-dependencies \
     && rm -rf /usr/src/arp-scan
 
+# PicoTTS - it has no specific version - commit should be taken from build.json
+ARG PICOTTS_HASH
+RUN apk add --no-cache popt \
+    && apk add --no-cache --virtual .build-dependencies automake autoconf libtool popt-dev build-base \
```
Comment (pvizeli) [inline Dockerfile:106]: each package on his own line

File: Dockerfile [hunk 13]
```diff
@@ -100,6 +100,21 @@ RUN apk add --no-cache \
     && apk del .build-dependencies \
     && rm -rf /usr/src/arp-scan
 
+# PicoTTS - it has no specific version - commit should be taken from build.json
+ARG PICOTTS_HASH
+RUN apk add --no-cache popt \
+    && apk add --no-cache --virtual .build-dependencies automake autoconf libtool popt-dev build-base \ 
+    && git clone https://github.com/naggety/picotts.git pico \
+    && cd pico \
```
Comment (pvizeli) [inline Dockerfile:108]: you can directly jump into cd pico/pico

File: Dockerfile [hunk 14]
```diff
@@ -100,6 +100,21 @@ RUN apk add --no-cache \
...
     && rm -rf /usr/src/arp-scan
 
+# PicoTTS - it has no specific version - commit should be taken from build.json
+ARG PICOTTS_HASH
+RUN apk add --no-cache popt \
+    && apk add --no-cache --virtual .build-dependencies automake autoconf libtool popt-dev build-base \ 
+    && git clone https://github.com/naggety/picotts.git pico \
+    && cd pico \
+    && git reset --hard ${PICOTTS_HASH} \
+    && cd pico || exit 1 # Sources and make files are in sub-dir ./pico, satisfy hadolint SC2164 \
```
Comment (pvizeli) [inline Dockerfile:110]: is not needed, see comment above

**Review by pvizeli — CHANGES_REQUESTED** [review]

**Comment by carstencodes** [comment]
> https://github.com/home-assistant/docker/pull/142/checks?check_run_id=1310419996#step:5:117 looks wrong. There is no cleanup

Yeah, looks like hadolint has been confused by the comment at the end of the line and now the step exits to early. I'll fix it tomorrow.

**Comment by homeassistant** [comment]
Hi @kngharv, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;docker#142).

Once you do that we will be able to review and accept this pull request.

Thanks!

File: Dockerfile (hunk 11 above)
Comment (carstencodes) [reply Dockerfile]: Done

File: Dockerfile (hunk 13 above)
Comment (carstencodes) [reply Dockerfile]: Done

File: Dockerfile (hunk 14 above)
Comment (carstencodes) [reply Dockerfile]: Done

File: Dockerfile (hunk 12 above)
Comment (carstencodes) [reply Dockerfile]: Done

**Comment by carstencodes** [comment]
@frenck @pvizeli Everything ok with this right now?

File: Dockerfile [hunk 15]
```diff
@@ -100,6 +100,26 @@ RUN apk add --no-cache \
...
+       automake \
+       autoconf \
+       libtool \
+       popt-dev \
+       build-base \ 
+    && git clone https://github.com/naggety/picotts.git pico \
+    && cd pico/pico \
+    && git reset --hard ${PICOTTS_HASH} \
+    && ./autogen.sh \
+    && ./configure \
```
Comment (pvizeli) [inline Dockerfile:117]:
```suggestion
    && ./configure \
        --disable-static \
```

File: build.json [hunk 16]
```diff
@@ -1,5 +1,8 @@
 {
   "image": "homeassistant/{arch}-homeassistant-base",
+  "args":{
+    "PICOTTS_HASH": "e3ba46009ee868911fa0b53db672a55f9cc13b1c"
+  },
```
Comment (pvizeli) [inline build.json:5]: please rebase and add it to exists args

**Review by pvizeli — CHANGES_REQUESTED** [review]
I think after that it should be fine

**Review by pvizeli — APPROVED** [review]

**Pull request merged**